iceoryx_posh/include/iceoryx_posh/internal/popo/used_chunk_list.hpp

```diff
@@ -21,161 +22,107 @@
-                m_synchronizer.clear(std::memory_order_release);
-                return true;
-            }
-            previous = current;
+            static_assert(sizeof(RelativePointerData) <= 8U, "The RelativePointerData size must not exceed 64 bit!");
+            static_assert(std::is_trivially_copyable<RelativePointerData>::value,
+                          "The RelativePointerData must be trivially copyable!");
         }
-        return false;
-    }
-
-    // only once from runtime context
-    void setup()
-    {
-        m_synchronizer.test_and_set(std::memory_order_acquire);
-        /// @todo maybe also call init here?
-    }
-
-    // from RouDi context once the applications walked the plank
-    void cleanup()
-    {
-        m_synchronizer.test_and_set(std::memory_order_acquire);
-
-        for (auto& data : m_data)
+        constexpr RelativePointerData(uint16_t segment, uint64_t offset) noexcept
+            : segment(segment)
+            , offset(offset & MAX_OFFSET)
         {
-            if (data != nullptr)
-            {
-                mepoo::SharedChunk{data};
-            }
+            cxx::Ensures(offset < MAX_OFFSET && "offset must not exceed MAX:OFFSET!");
         }
 
-        init(); // just to save us from the future self
-    }
+        uint16_t segment : 16;
+        uint64_t offset : 48;
 
-  private:
-    void init()
-    {
-        // build list
-        for (uint32_t i = 0u; i < Size; ++i)
+        void reset() noexcept
         {
-            m_list[i] = i + 1u;
+            constexpr RelativePointerData NULLPTR_EQUIVALENT{};
+            *this = NULLPTR_EQUIVALENT;
         }
 
-        if (Size > 0)
-        {
-            m_list[Size - 1u] = InvalidIndex; // just to save us from the future self
-        }
-        else
+        bool isNullptr() const noexcept
         {
-            m_list[0u] = InvalidIndex;
+            return segment == MAX_SEGMENT && offset == MAX_OFFSET;
         }
 
+        static constexpr uint64_t MAX_SEGMENT{std::numeric_limits<uint16_t>::max()};
+        static constexpr uint64_t MAX_OFFSET{(1ULL << 48) - 1};
+    };
 
-        m_usedListHead = InvalidIndex;
-        m_freeListHead = 0u;
-
-        // clear data
-        for (auto& data : m_data)
-        {
-            data = nullptr;
-        }
-
-        m_synchronizer.clear(std::memory_order_release);
-    }
-
-    bool freeSpaceInList()
-    {
-        return (m_freeListHead != InvalidIndex);
-    }
-
-
-  private:
-    static constexpr uint32_t InvalidIndex{Size};
+    using DataElement_t = RelativePointerData;
+    static constexpr DataElement_t DATA_ELEMENT_NULLPTR{};
 
   private:
     std::atomic_flag m_synchronizer = ATOMIC_FLAG_INIT;
-    uint32_t m_usedListHead{InvalidIndex};
+    uint32_t m_usedListHead{INVALID_INDEX};
     uint32_t m_freeListHead{0u};
-    std::array<uint32_t, Size> m_list;
-    std::array<rp::RelativePointer<mepoo::ChunkManagement>, Size> m_data;
+    uint32_t m_listNodes[Capacity];
```

Comment:
I find the naming misleading. It is not a node, it is an index to the m_listData array. Its elements encode the next index in the (implicit) list(s).

Reply:
I will change it to `m_listIndices`